Adding Convergence Plots

atharva-2001 opened the pull request:
<!--- Provide a general summary of your changes in the title above -->
This PR adds convergence plots to TARDIS. 

**Description**
<!--- Describe your changes in detail -->
Convergence plots will allow the user to see the convergence of the simulation in plots and study the state of the simulation at different iterations. Convergence plots are plotted by default as the simulation runs. One can configure the plots by providing arguments in the `run_tardis` function. This includes overriding the default plotly layout of the graphs and options to change color scale and show or not show the plots.

**Motivation and context**
<!--- Why is this change required? What problem does it solve? Link issues here -->

**How has this been tested?**
- [X] Testing pipeline.
- [ ] Other. <!--- please describe how you tested your changes, `pytest` flags used, etc. -->

**Examples**
<!-- If appropiate, link notebooks, screenshots and other demo stuff -->
[Link](https://atharva-2001.github.io/tardis/branch/convergence_plots/io/visualization/convergence_plot) to the convergence plots in the documentation.
[Link](https://nbviewer.jupyter.org/github/atharva-2001/tardis/blob/convergence_plots/docs/io/visualization/convergence_plot.ipynb ) to the notebook.
[Link](https://atharva-2001.github.io/tardis/branch/convergence_plots/api/tardis.visualization.tools.convergence_plot.html) to the API
![cplots-new2](https://user-images.githubusercontent.com/55894364/122961172-c5ffc080-d3a1-11eb-9fc7-9c48e2e15154.gif)

Convergence Plots along with progress bars
![Peek 2021-07-09 18-15](https://user-images.githubusercontent.com/55894364/125080064-f9e91e80-e0e1-11eb-8edf-983e02dfda67.gif)

**Type of change**
<!--- Put an `x` in all the boxes that apply -->
- [ ] Bug fix. <!-- non-breaking change which fixes an issue -->
- [X] New feature. <!-- non-breaking change which adds functionality -->
- [ ] Breaking change. <!-- fix or feature that would cause existing functionality to not work as expected -->
- [ ] None of the above. <!-- please describe -->

**Checklist**
<!--- Put an `x` in all the boxes that apply -->
- [X] My change requires a change to the documentation.
    - [X] I have updated the documentation accordingly.
    - [X] (optional) I have built the documentation on my fork following [the instructions](https://tardis-sn.github.io/tardis/development/documentation_preview.html).
- [X] I have assigned and requested two reviewers for this pull request.

MarkMageeAstro commented:
This looks like a great start. I have a few stylistic comments:

Plasma updates
1. Can you swap the plot round so that temperature is on the left? There are no labels for temperature or velocity, is this the latex rendering issue?
2. I think the first example you have with iterations and then just the numbers looks the best. It's hard to tell exactly what's going on in the last one because the lines aren't connected and that also makes it difficult to see the colours on the right hand side
3. It would be best to change the display formatting and add units. The full temperature should be printed out, not just rounded to the nearest 1000. For example, use 10,000 K and not 10k K. Similarly, the x-axis should be in km/s but at the minute is in cm/s
4. Personally, I think a different colour scheme would be good, like rainbow or jet. If we stick with this one then I would at least swap it round so that the lighter yellow is the last iteration.


Temperature plots
1. We need units here too. Luminosity should be in erg / s
2. A bar chart isn't really the best way to present this data, so I would remove it. The line plot is much better. It should also have a legend to say what each colour is and to say what the horizontal line is. I would use some more contrasting colours as well
3. I think the labels are the wrong way round. In the line plot, the final emitted luminosity is listed as 4e42, but the simulation output says this is the luminosity absorbed. 
4. Might also be nice to show a residual difference between the emitted and requested luminosities for each iteration. This can just be the difference between the two given as a percentage of the luminosity requested or something like that. Could be a subpanel to the main plot, just below

jaladh-singhal reviewed (commented):
Apparently, plots didn't show up for me when running simulation - do I have to enable them by passing option? (I thought we were keeping plots by default on and logging turned off)

atharva-2001 commented:
I'm sorry, I had done that to see if I was able to pass the tests. Could you please try once more? The plots are on now by default.

atharva-2001 commented:
Thanks, @MarkMageeAstro! 
I'm sorry for not changing the `t_inner` damping constant. I changed it now.  I also changed the height of the plots and removed the whitespace. Could you please tell me if that looks good to you? 
The `requested_luminosity` line, among other lines in the luminosity plot, do not show marker points by default. 
I just changed it to demonstrate that the default configuration can be overridden. 
![image](https://user-images.githubusercontent.com/55894364/122963985-6e168900-d3a4-11eb-9f40-0263d2b4e1c4.png)
I'm afraid the link to the notebook in the documentation is generated automatically due to code in `docs/conf.py`. The link will be fixed once the PR is merged. 
I guess the less number of ticks was due to the small width of the plot. It's showing 4 ticks for me now. Should I increase the number of ticks?  Do you mean adding ticks like in [this](https://stackoverflow.com/questions/38064175/major-and-minor-tickmarks-with-plotly)?

atharva-2001 commented:
I'm afraid there isn't a direct way to add minor ticks to plotly charts, but there are workarounds. We would have to specify the names of the ticks or, `ticktext`, as well as the values of the ticks or, `tickvals`. Leaving the `ticktext` of some of the ticks as an empty string would create minor ticks. But we would have to automate finding `tickvals` and `ticktext` according to the simulation result. I'm still working on it, though.

jaladh-singhal commented on tardis/simulation/base.py:
```diff
@@ -162,6 +167,16 @@ def __init__(
                 f"- input is {convergence_strategy.type}"
             )
 
+        if cplots_kwargs["show_plots"] is not False:
```
```suggestion
        if cplots_kwargs["show_plots"]:
```
or ` if cplots_kwargs["show_plots"] is True:`

Why not keep it simple :smile:

jaladh-singhal commented on tardis/simulation/base.py:
```diff
@@ -553,6 +605,17 @@ def from_config(
+            "show_plots",
+            "export_cplots",
+        ]
+        cplots_kwargs = defaultdict(dict)
```
Is this for creating an empty dictionary?

If yes, then why not do `= {}`?

jaladh-singhal commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,360 @@
+from traitlets import TraitError
+
+
+def transition_colors(name="jet", iterations=20):
```
I don't understand why you're using a default value of iterations, shouldn't it always come from config file?

atharva-2001 replied:
I'm sorry, my bad, the number of iterations always come from the config file, I just forgot to remove `iterations=20`

jaladh-singhal commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,360 @@
+        else:
+            self.colorscale = transition_colors(iterations=self.iterations)
+
+        if "plasma_plot_config" in kwargs:
+            if kwargs["plasma_plot_config"] != {}:
```
Better to `and` the both. Also `if x != {}` is equivalent to `if x` - we're checking truthy/falsy values

jaladh-singhal commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,360 @@
+            if kwargs["plasma_plot_config"] != {}:
+                self.plasma_plot_config = kwargs["plasma_plot_config"]
+
+        if "luminosity_plot_config" in kwargs:
+            if kwargs["luminosity_plot_config"] != {}:
```
Same as above - and both

jaladh-singhal commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,360 @@
+            ),
+            yaxis2=dict(
+                exponentformat="e",
+                title=r"$\mbox{Luminosity}~(erg~sec^{-1})$",
```
```suggestion
                title=r"$\mbox{Luminosity}~(erg~s^{-1})$",
```

jaladh-singhal commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,360 @@
+        Creates an empty luminosity plot.
+        The default layout can be overridden by passing luminosity_plot_config dictionary in the run_tardis function.
+        """
+        line_colors = ["orangered", "lightseagreen", "indigo"]
```
Is it documented how can one override these colors?

jamesgillanders replied:
are these colours colour-blind friendly? the defaults should be suitable for all users. I think red/green is a no-go, and here we have orangered/green?

jamesgillanders replied:
^

atharva-2001 replied:
I'm afraid the colors are not color blind friendly. I will change that soon. These colors can be changed by passing a list of them or the name of the color in the `luminosity_colorscale` option. 
Something like this:
```py
sim = run_tardis(
    "tardis_example.yml",
    luminosity_colorscale = "viridis", 
)
```
Or this:
```py
sim = run_tardis(
    "tardis_example.yml",
    luminosity_colorscale = [
        'rgb(229, 134, 6)',
        'rgb(93, 105, 177)',
        'rgb(82, 188, 163)',
        'rgb(153, 201, 69)',
        'rgb(204, 97, 176)',
    ],
)
```

atharva-2001 replied:
Functionality added in https://github.com/tardis-sn/tardis/commit/38bcb465adf12322816f3194badb0fd3ca525a11, added documentation in https://github.com/tardis-sn/tardis/commit/d26efd54ea35a0cc121ab96612e0df1d8bd8f94e

jaladh-singhal commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,360 @@
+
+    def build(self, display_plot=True):
+        """
+        Calls the create_plasma_plot and the create_luminosity_plot to build plots.
```
A docstring of function should never be concerned with what functions it calls - think about it, a user reading your don't know inner mechanics of your code so just writing what functions this function calls isn't helpful. Instead write what it does including the what the functions it call does.

For e.g. something like: "Create both plasma plot and luminosity plot and display them together."

jaladh-singhal commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,360 @@
+
+    def update_plasma_plots(self):
+        """
+        Updates the plasma plots using the data collected using the fetch_data function.
```
Again no need to tell `..using the fetch_data function` instead explain what it does

jaladh-singhal commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,360 @@
+
+    def update_luminosity_plot(self):
+        """
+        Updates the plasma plots using the data collected using the fetch_data function.
```
Fix this too

jaladh-singhal commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,360 @@
+
+        self.current_iteration += 1
+        if export_cplots:
+            with suppress(TraitError):
```
Good that you supressed only TraitError. Did you figure out that why this error arises (becuase supressing should be our last resort)?

atharva-2001 replied:
The TraitError arises because it was not accepting None as a return value, but `fig.show()` returns None.
```
TraitError: The 'children' trait of a VBox instance contains an Instance of a TypedTuple which expected a Widget, not the NoneType None.
```
No error arises when displaying the plots naturally:
```py 
display(widgets.VBox([go.FigureWidget(), go.FigureWidget()])) # No error
```
But only when one uses `fig.show()` to display the plots (we are doing this only when we are exporting plots)
```py
display(widgets.VBox([go.FigureWidget().show(), go.FigureWidget().show()])) # TraitError
```

jaladh-singhal commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,360 @@
+        """
+        Calls functions used to build and update convergence plots.
+        """
+        if self.iterable_data != {}:
+            if self.current_iteration == 1:
```
`and` both conditions

atharva-2001 replied:
I'm afraid the second `if` statement builds the empty plots and is only executed if it's the first iteration. Using `and` won't allow building the plots only once.

jaladh-singhal commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,360 @@
+
+    def update(self, export_cplots=False, last=False):
+        """
+        Calls functions used to build and update convergence plots.
```
Again same thing: something like "Build and update convergence plots for each iteration"

jaladh-singhal commented on tardis/visualization/widgets/shell_info.py:
```diff
@@ -4,7 +4,7 @@
     species_tuple_to_string,
 )
-from tardis.simulation import Simulation
+
```
Was Simulation not being used in the module?

atharva-2001 replied:
I'm afraid not

jaladh-singhal commented on tardis/visualization/tools/tests/test_convergence_plot.py:
```diff
@@ -0,0 +1,43 @@
+def test_build(convergence_class):
+    convergence_class.build(display_plot=False)
+    assert type(convergence_class.plasma_plot) == go.FigureWidget
+    assert type(convergence_class.luminosity_plot) == go.FigureWidget
```
This is a good start for tests - you might also want to compare values besides types?

jaladh-singhal reviewed (changes requested):
This is an excellent work @atharva-2001 - good code structuring.

There are few comments and questions I've left. The main thing is to write better docstrings - descriptions of several are not very helpful and in some, you didn't documented parameters. Also I think there's need of adding code comments (which describe what a particular part of your code is doing) - they're almost negligible. Please give time to read SDEC plot and line info widget modules, you'll understand both things!

MarkMageeAstro commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,389 @@
+        fig = fig.update_layout(
+            xaxis={
+                "tickformat": "g",
+                "title": r"$Shell~~Velocity$",
```
Change "Shell Velocity" to just "Velocity" and add the units

atharva-2001 replied:
![image](https://user-images.githubusercontent.com/55894364/123802452-f21ec280-d908-11eb-94eb-a7493081013e.png)
This is what I have done. Is this what you had in mind?

MarkMageeAstro replied:
Yes, this is much better

MarkMageeAstro commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,389 @@
+            },
+            xaxis2={
+                "tickformat": "g",
+                "title": r"$Shell~~Velocity$",
```
Same here

MarkMageeAstro commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,389 @@
+        )
+
+        fig.add_scatter(
+            name="Inner<br>Boundary<br>Temperature",
```
Can you put the temperature legend first, since that is the first plot

MarkMageeAstro commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,389 @@
+                "tickformat": "g",
+                "title": r"$W$",
+                "tickmode": "auto",
+                "nticks": 15,
```
Is it possible to add some of these things, like nticks, height, etc., as kwargs? Or some method of having this be user configurable

atharva-2001 replied:
The `height` and many other layout properties can be changed using the `plasma_plot_config` and the `luminosity_plot_config` options, but I'm afraid I didn't get the desired results using that method only for `nticks`. Apart from `nticks`, I believe most of the properties can be overridden. Setting `nticks` to say 10 will sometimes not create 10 `nticks`, but it does increase them, arbitrarily. I am working on it and will create an issue in plotly if I am able to recreate it.

MarkMageeAstro commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,389 @@
+        Updates plasma convergence plots every iteration.
+        """
+        x = self.iterable_data["velocity"].value.tolist()
+        x = [item / 100000 for item in x]
```
Should these be a unit conversion instead?

MarkMageeAstro reviewed (changes requested):
Another nice PR, @atharva-2001. I've left a few specific comments. Another general request from me would be to add some more code comments. There are some docstrings, but having comments alongside some lines of code would be very useful too.

jamesgillanders commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,388 @@
+            legend_title_text="Iterations",
+            margin=dict(l=10, r=135, b=25, t=25, pad=0),
+        )
+
```
Label units should not be in italics. Also, the large capitals in T_rad and W should be italicised, but the subscripts should not be (same logic applies to all parameters, e.g. T_inner later).

jamesgillanders commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,388 @@
+            hoverlabel_font_color="white",
+            margin=dict(b=25, t=25, pad=0),
+        )
+
```
again, remove italicisation of units here

jamesgillanders commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,388 @@
+            customdata=customdata,
+            hovertemplate="<b>Y</b>: %{y:.2f} at <b>X</b> = %{x:,.0f}%{customdata}",
+        )
+
```
Is this where the values for Luminosity are generated for the hover box? If so, can you increase the number of significant figures output by default? 1 decimal place truncates a lot of useful info; e.g. see screenshot here. I would suggest 3 decimal places


![Screenshot 2021-06-29 at 15 34 11](https://user-images.githubusercontent.com/47385282/123816754-78c6a600-d8ef-11eb-8b1a-45711dbd630f.png)

jamesgillanders replied:
this applies to the other plots for T_rad too

jamesgillanders commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,388 @@
+            + "Absorbed Luminosity: "
+            + f'{self.value_data["Requested"][-1]:.2g}'
+        ]
+
```
Not too sure where the legend stuff is defined, but why is there a Luminosity header for all the lines? This doesn't work if the first line in there is for temperature! Either find a way to separate the temp. line out and then have the Luminosity header for the rest of the lines (and if you do this, remove the luminosity second words from all those labels - e.g. 'emitted luminosity' would become 'emitted') or just remove 
the header. I recall seeing you were finding it difficult to produce 2 separate legends, so perhaps just remove the header.

![Screenshot 2021-06-29 at 15 40 58](https://user-images.githubusercontent.com/47385282/123817859-6a2cbe80-d8f0-11eb-8f2e-64e01e7ac71f.png)

atharva-2001 replied:
done in https://github.com/tardis-sn/tardis/commit/b652101a3ddb66107fe4b73019f63f7834a2a4d7

jamesgillanders reviewed (commented):
Code and plots looks good @atharva-2001 - good job! Most of the major things have already been pointed out by the other reviewers, so here are my minor nitpicky things. If/when you get these resolved, I think we'd bee ready to think about merging!

atharva-2001 commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,388 @@
+        iteration number
+    """
+
+    def __init__(self, iterations, **kwargs):
```
done in https://github.com/tardis-sn/tardis/commit/d26efd54ea35a0cc121ab96612e0df1d8bd8f94e

MarkMageeAstro commented:
I think this may need rebased because the analysis/opacities script is showing up as a new file added here.

jaladh-singhal reviewed (changes requested):
This looks a lot better than before - tests are good. You've fixed docstrings but there are still some problems. And I can see you've added code comments but they're still missing from some modules.

jaladh-singhal reviewed (commented):
Excellent work @atharva-2001 - most of the docstrings look very good now! Thanks for the changes.

jaladh-singhal commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,439 @@
+                display(
+                    widgets.VBox(
+                        [
+                            self.plasma_plot.show(
+                                renderer="notebook_connected"
+                            ),
+                            self.t_inner_luminosities_plot.show(
+                                renderer="notebook_connected"
+                            ),
```
Is it possible to not call show instead just return FigureWidget - I think then display will just show it and you won't need to supress trait error?

atharva-2001 replied:
Actually, this exact thing happens in the `build` function. The plots show up locally without having to suppress any errors, but in order to display them on platforms like nbviewer, we need to have the [`notebook_connected`](https://plotly.com/python/renderers/) renderer. Unfortunately `fig.show()` returns None which causes the error we need to suppress. 

If it would have worked, we would have not needed the `export_cplots` option but would have seen the plots two times online as we see now locally, if we had run the simulation with `export_cplots=True`. One due to the build function and the other due to the `notebook_connected` renderer. The build function fails to display the plots online actually.

jaladh-singhal replied:
OK - thanks for explanation!

jaladh-singhal commented on docs/io/visualization/convergence_plot.ipynb:
```diff
@@ -0,0 +1,846 @@
+{
```
<p>Highlight the fact that they are stored in <code>cplots</code>  attribute of simulation object <code>sim</code> . And also phrase it out that we have two plots - plasma and t_inner &amp; luminosites within it</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1636/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

jaladh-singhal replied:
@atharva-2001 I believe only this is left?

atharva-2001 replied:
I guess I fixed this in https://github.com/tardis-sn/tardis/pull/1636/commits/4401372fd30d2c9846db172d9f546e200e26eca0
Could you please check if it's up to the mark? 
[documentation](https://atharva-2001.github.io/tardis/branch/convergence_plots/io/visualization/convergence_plot)

jaladh-singhal commented on docs/io/visualization/convergence_plot.ipynb:
```diff
@@ -0,0 +1,846 @@
+{
```
<p>Is changing the layout here necessary? I think this cell should only focus on show them up since you're already describing how to change layout below</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1636/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

jaladh-singhal commented on docs/io/visualization/convergence_plot.ipynb:
```diff
@@ -0,0 +1,846 @@
+{
```
<p>This is excellent, as well as description of changing colors section</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1636/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

atharva-2001 replied:
Thank you!!

atharva-2001 commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,439 @@
+        Dictionary used to override default plot properties of plasma plots.
+    t_inner_luminosities_config : dict, optional
+        Dictionary used to override default plot properties of the inner boundary temperature and luminosity plots.
+    plasma_cmap : str, default: 'jet', optional
```
I'm afraid I think what I did on L23 was incorrect? Please correct me if I am wrong. The [examples](https://numpydoc.readthedocs.io/en/latest/format.html#parameters) in the `NumPy` style guide have mentioned default values next to types.  [This](https://python-sprints.github.io/pandas/guide/pandas_docstring.html#section-3-parameters) example also has default values next to types.

jaladh-singhal replied:
OK but do that everywhere in your module. Each optional parameter must contain `default: xyz` in your parameter type and description should explain what that does.

jaladh-singhal commented on tardis/visualization/tools/convergence_plot.py:
```diff
@@ -0,0 +1,440 @@
+
+        Parameters
+        ----------
+        display_plot : bool, optional
+            Displays empty plots. Defaults to True.
```
`default: True` should go in type instead of description - like you did everwhere else

jamesgillanders reviewed (approved):
This all looks great @atharva-2001! All requested changes seem to have been addressed, and the code itself works really well. Good job!

jaladh-singhal reviewed (approved):
Brilliant work! Thank you for addressing all of my nitpicky comments :smile:

jaladh-singhal commented:
@atharva-2001 this needs rebasing now that #1723 is merged
